### ISSUE TITLE
Flatten modrules

### ISSUE THREAD

**Issue by sprunk** [beyond-all-reason/spring#1260]
The subtables don't bring any value and just make it difficult to work with the file at all ends. Keep the subtable entries as backup for now though.

**Comment by sprunk** [comment]
Ideally there would be a flat `Game.modRules.*` table exposed to Lua as well which would perfectly mirror the one accepted in `modrules.lua`. Right now modrules are mixed together with other random stuff directly in `Game`.

**Comment by sprunk** [comment]
More or less, but there's a caveat that I want to use this oportunity to change the name and/or format (while keeping the current one in the legacy subtables). See #939 and #940

For example the current way is:
```lua
reclaim = {
  unitEfficiency = 0.123, -- first resource only
  unitEnergyCostFactor = 0.456, -- second resource only, always referred to as "energy"
  featureEnergyCostFactor = 0.789, -- second resource only
},
repair = {
  energyCostFactor = 0.876 -- second resource only
},
capture = {
  energyCostFactor = 0,
},
```
while the equivalent flat version would look like (in an example game with 3 fancy resources):
```lua
reclaimUnitEfficiency = { requisition = 0.123 }, -- or {0.123, 0, 0}
reclaimUnitCostFactor = { requisition = 0, blood = 0.456, ore = 0 },
reclaimFeatureCostFactor = { 0, 0.789, 0 },
repairCostFactor = { requisition = 0.876, blood = 0, ore = 0 },
captureCostFactor = { requisition = 0, blood = 0, ore = 0.555 },
```

**Comment by marcushutchings** [comment]
I see. How easy is it to support arbitrary resource names like that?

**Comment by sprunk** [comment]
Should be trivial since these exist:
https://github.com/beyond-all-reason/RecoilEngine/blob/41c58746c30843fe946332f1a2f868f27c430149/rts/Sim/Misc/ResourceHandler.cpp#L57
https://github.com/beyond-all-reason/RecoilEngine/blob/41c58746c30843fe946332f1a2f868f27c430149/rts/Sim/Misc/ResourceHandler.cpp#L68
https://github.com/beyond-all-reason/RecoilEngine/blob/41c58746c30843fe946332f1a2f868f27c430149/rts/Sim/Misc/Resource.h#L140

So the end state looks like this:
* modrule parser reads a list of resources as one of the first steps. For example
```lua
resources = { "metal", "requisiton", "blood", "ore" },
```
* modrule parser does `resourceHandler->AddResource(bla)` a bunch of times, which contains the resource names.
* afterwards, when reading things that deal with resources (such as `repairCostFactor`), it does
```cpp
for (int i = 0; i < SResourcePack::MAX_RESOURCES; ++i) {
	const auto resourceDef = resourceHandler->GetResource(i);
	if (resourceDef == nullptr)
		break; // no more resources defined
	resourcePack[i] = resourcesTable.GetFloat(i,                 resourcePack[i]);
	resourcePack[i] = resourcesTable.GetFloat(resourceDef->name, resourcePack[i]);
}
```
* the same applies to everywhere else, not just modrules (unit defs, LuaUI/LuaRules resource API, etc).
* modrules and the resource handler don't seem to be subject to CREG at the moment but fundamentally resource defs are just a vector of numbers and strings so it should be trivial to add them.

**Comment by saurtron** [comment]
Not a problem for me, but I get the impression flattening the modrules is controversial, see [this comment[(https://github.com/beyond-all-reason/spring/pull/1802#issuecomment-2525240425) by @lhog.

Would be good to have a final say/decision over than and then get done with it.

**Comment by sprunk** [comment]
I have an implementation mostly stashed and I wouldn't expect to remove back-compat anytime soon either way.

**Comment by saurtron** [comment]
> I have an implementation mostly stashed and I wouldn't expect to remove back-compat anytime soon either way.

Yeah, ok actually looking at that conversation looks like @lhog is ok with it reading further below. Just wanted to make sure sorry.
